src/contracts/scanner.rs

```diff
@@ -0,0 +1,58 @@
+use ethers::providers::Middleware;
+use ethers::types::{Address, BlockNumber, Filter, FilterBlockOption, Log, Topic, ValueOrArray};
+
+pub struct BlockScanner<T> {
+    read_provider: T,
+    current_block: u64,
+    window_size:   u64,
+}
+
+impl<T> BlockScanner<T>
+where
+    T: Middleware,
+    <T as Middleware>::Error: 'static,
+{
+    pub async fn new_latest(read_provider: T, window_size: u64) -> anyhow::Result<Self> {
+        let latest_block = read_provider.get_block_number().await?;
+
+        Ok(Self {
+            read_provider,
+            current_block: latest_block.as_u64(),
+            window_size,
+        })
+    }
```

Comment:
If we ever want to start from an earlier block than the latest, consider changing this to 

```rust
pub async fn new_latest(read_provider: T, window_size: u64, block_number: Option<BlockNumber>) -> anyhow::Result<Self> {
    let block_number = if let Some(block_number){
        block_number
    } else{
        read_provider.get_block_number().await?
    };

    Ok(Self {
        read_provider,
        current_block: block_number.as_u64(),
        window_size,
    })
}
```

Reply:
If we want to start from a specific block number I think I'd want a new `new` function which accepts specific block number. Which would make this function infallible and non-const